### PR TITLE
Add Groundpound powerup

### PR DESCRIPTION
**Pull request by legokidlogan**

### CODE REVIEW

File: lua/effects/powerups_groundpound_shockwave_huge/init.lua [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    self.Emitter = ParticleEmitter( self.Pos ) -- Emitter must be there so you don't get an error
+
+    sound.Play( "ambient/explosions/explode_" .. math.random( 1, 4 ) .. ".wav", self.Pos, 100, 100 )
+
+    self:Dust()
+
+    self.Emitter:Finish()
+end
+
+function EFFECT:Dust()
```
Comment (sarahsturgeon) [inline lua/effects/powerups_groundpound_shockwave_huge/init.lua:16]: from experience, this many effects and emitters can kill weaker pcs

File: lua/powerups/server/groundpound.moon [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+
+            -- Apply knockback to victims
+            knockback = @knockbackMult * damageInfo\GetDamage!
+            dir = victim\WorldSpaceCenter! - @owner\GetPos!
+
+            -- Force the direction to have a significant upwards angle
+            -- Otherwise, it's a coin flip between barely any movement and way too much, because of Source shenanigans
+            ang = dir\Angle!
+            pitch = -math.Rand 40, 60
+            dir = ( Angle pitch, ang.yaw, 0 )\Forward!
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]: you can just
```suggestion
            dir = Angle(pitch, ang.yaw, 0)\Forward!
```

File: lua/powerups/server/groundpound.moon [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+            @UsesRemaining -= 1
+            @owner\ChatPrint "You have #{math.max @UsesRemaining, 0} Groundpound uses remaining"
+
+            util.ScreenShake ownersPos, speedZeroToOne * 40, 4, 2, 1000 -- big shake close
+            util.ScreenShake ownersPos, 1, 5, 5, damage * 2 -- small shake far away
+
+            with @owner
+                -- double damage for whatever we directly land on
+                groundEntity = @owner\GetGroundEntity!
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon:163]:
```suggestion
                groundEntity = \GetGroundEntity!
```

File: lua/powerups/server/groundpound.moon [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+                    effScale = speedZeroToOne * 0.5
+                    effScale = effScale + 0.2
+                    with effDat
+                        \SetOrigin ownersPos
+                        \SetNormal UP_VECTOR
+                        \SetScale effScale
+                        util.Effect "powerups_groundpound_shockwave", effDat
+
+            if @UsesRemaining < 1
+                @Remove!
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]:
```suggestion
            @Remove! unless @UsesRemaining > 0
```

File: lua/powerups/server/groundpound.moon [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+            util.BlastDamage @damageInflictor, @owner, ownersPos, radius, damage
+
+            @UsesRemaining -= 1
+            @owner\ChatPrint "You have #{math.max @UsesRemaining, 0} Groundpound uses remaining"
+
+            util.ScreenShake ownersPos, speedZeroToOne * 40, 4, 2, 1000 -- big shake close
+            util.ScreenShake ownersPos, 1, 5, 5, damage * 2 -- small shake far away
+
+            with @owner
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon:161]: This is a big hook function - separate some stuff out into helper methods?

File: lua/powerups/server/groundpound.moon [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+            cantFastFall = checkCantFastFall!
+
+            if cantFastFall
+                if @fastFalling
+                    @fastFalling = false
+                    @upToSpeedSound = false
+                    @upToTerminalSpeedSound = false
+                    @fallSound\ChangeVolume 0, FALL_SOUND_FADE_OUT
+
+                return
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]:
```suggestion
            if cantFastFall
                return unless @fastFalling

                @fastFalling = false
                @upToSpeedSound = false
                @upToTerminalSpeedSound = false
                @fallSound\ChangeVolume 0, FALL_SOUND_FADE_OUT
```

File: lua/powerups/server/groundpound.moon [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+            if not @fastFalling
+                @fastFalling = true
+                @fallSound\ChangeVolume 1, FALL_SOUND_FADE_IN
+
+                now = CurTime!
+
+                if @nextCrouchSound <= now -- sound when starting the groundpound
+                    @owner\EmitSound "ambient/machines/thumper_top.wav", 78, 110, 1
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]:
```suggestion
                    owner\EmitSound "ambient/machines/thumper_top.wav", 78, 110, 1
```

File: lua/powerups/server/groundpound.moon [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+            -- Downwards acceleration
+            change = -@accel * dt
+            vel.z = vel.z + change
+            velToAdd = Vector 0, 0, change
+
+            speedFrac = math.max -vel.z / TERMINAL_VELOCITY, 0 -- 0 to 1 based on how close to terminal velocity we are
+            fallSoundPitch = Lerp speedFrac, 100, 200
+            fallSoundLevel = Lerp speedFrac, 75, 150
+
+            @owner\SetVelocity velToAdd
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon:315]:
```suggestion
            owner\SetVelocity velToAdd
```

File: lua/powerups/server/groundpound.moon [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+            @fallSound\ChangePitch fallSoundPitch, dt
+            @fallSound\SetSoundLevel fallSoundLevel
+
+            return if vel.z > -@minSpeed
+
+            if not @upToSpeedSound -- sound when starting to deal damage
+                @upToSpeedSound = true
+                if @nextUpToSpeedSound < CurTime! -- block spamming this sound
+                    @owner\EmitSound "weapons/mortar/mortar_shell_incomming1.wav", 120, 100, 0.5
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]:
```suggestion
                    owner\EmitSound "weapons/mortar/mortar_shell_incomming1.wav", 120, 100, 0.5
```

File: lua/powerups/server/groundpound.moon [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+                
+                @nextUpToSpeedSound = CurTime! + 15
+            if vel.z <= -TERMINAL_VELOCITY and not @upToTerminalSpeedSound -- sound when hitting terminal velocity
+                @upToTerminalSpeedSound = true
+                if @nextUpToTerminalSpeedSound < CurTime!
+                    filter = RecipientFilter!
+                    filter\AddAllPlayers!
+                    @owner\EmitSound "weapons/mortar/mortar_shell_incomming1.wav", 150, 60, 0.5, CHAN_AUTO, 0, 0, filter
```
Comment (sarahsturgeon) [inline lua/powerups/server/groundpound.moon]:
```suggestion
                    owner\EmitSound "weapons/mortar/mortar_shell_incomming1.wav", 150, 60, 0.5, CHAN_AUTO, 0, 0, filter
```

**Review by sarahsturgeon — CHANGES_REQUESTED** [review]
I'd like some minor code rework in the big methods. Also, I'd like to make sure this is tested on potato PCs before we release it

File: lua/effects/powerups_groundpound_shockwave_huge/init.lua (hunk 1 above)
Comment (legokidlogan) [reply lua/effects/powerups_groundpound_shockwave_huge/init.lua:16]: It's the same as `m9k_gdcw_cinematicboom` (which gets used by a variety of things in m9k) without the 'Debrizzle' particles, which increase in quantity with `Scale * Magnitude`.

However, m9k usually leaves Scale (which affects the other particle counts) in the range of 1-2, only using 4 for the ammo crate explosions. I'll take a stab ad reducing the particle amount.

File: lua/powerups/server/groundpound.moon (hunk 6 above)
Comment (legokidlogan) [reply lua/powerups/server/groundpound.moon]: That would change the logic. I want it to always return if `cantFastFall` is true, but *also* change those vars if it was previously fast falling.

File: lua/effects/powerups_groundpound_shockwave_huge/init.lua (hunk 1 above)
Comment (sarahsturgeon) [reply lua/effects/powerups_groundpound_shockwave_huge/init.lua:16]: Alright. If you can't reasonably test its performance on other setups, we should just keep a close eye out for anyone complaining about performance while it's being used 👍

**Pull request merged**